editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -0,0 +1,359 @@
+      }
+    } else {
+      if (isProjectContentFile(secondContents)) {
+        changesToProcess.push(deletePathChange(fullPath, true))
+        changesToProcess.push(writeProjectFileChange(fullPath, secondContents.content))
+      } else {
+        // Do nothing, both sides are a directory.
+      }
+    }
+
+    return true
+  }
+
+  function onElement(
+    fullPath: string,
+    firstContents: ProjectContentsTree | null,
+    secondContents: ProjectContentsTree | null,
+  ): boolean {
+    if (firstContents == null) {
+      if (secondContents == null) {
+        // Do nothing, nothing exists.
+        return false
+      } else {
+        changesToProcess.push(
+          writeProjectFileChange(fullPath, getProjectFileFromTree(secondContents)),
+        )
+        return true
+      }
+    } else {
+      if (secondContents == null) {
+        changesToProcess.push(deletePathChange(fullPath, true))
+        return false
+      } else {
+        if (firstContents === secondContents) {
+          // Same value, stop here.
+          return false
+        } else {
+          return applyChanges(fullPath, firstContents, secondContents)
+        }
+      }
+    }
+  }
+  if (isBrowserEnvironment && isFeatureEnabled('Collaboration')) {
+    if (oldContents != newContents) {
+      zipContentsTree(oldContents, newContents, onElement)
+    }
+  }
+
+  return changesToProcess
+}
+
+function applyFileChangeToMap(
+  change: ProjectFileChange,
+  projectContentsMap: CollaborativeEditingSupportSession['projectContents'],
+  mergeDoc: Y.Doc,
+): void {
+  switch (change.type) {
+    case 'DELETE_PATH':
+      const keyIterator = projectContentsMap.keys()
+      let keyIteratorValue = keyIterator.next()
+      while (!keyIteratorValue.done) {
+        const key = keyIteratorValue.value
+        if (
+          key === change.fullPath ||
+          (change.recursive && key.startsWith(`${change.fullPath}/`))
+        ) {
+          projectContentsMap.delete(key)
+        }
+      }
+      break
+    case 'WRITE_PROJECT_FILE':
+      if (
+        change.projectFile.type === 'TEXT_FILE' &&
+        change.projectFile.fileContents.parsed.type === 'PARSE_SUCCESS'
+      ) {
+        const parsedPart = change.projectFile.fileContents.parsed
+        let collabFile: CollabTextFile
+        if (projectContentsMap.has(change.fullPath)) {
+          collabFile = projectContentsMap.get(change.fullPath)!
+          // TODO Handle not text type
+        } else {
+          collabFile = new Y.Map()
+          collabFile.set('type', 'TEXT_FILE')
+          const topLevelElementsArray = new Y.Array<TopLevelElement>()
+          collabFile.set('topLevelElements', topLevelElementsArray)
+        }
+
+        mergeDoc.transact(() => {
+          if (!projectContentsMap.has(change.fullPath)) {
+            projectContentsMap.set(change.fullPath, collabFile)
+          }
+          synchroniseParseSuccessToCollabFile(parsedPart, collabFile)
```

Comment:
does this need to run in the case when collabFile was set to a `new Y.Map()`? if not, you could move the mergeDoc.transact to inside the two branches of the if-else statement, and remove the variable `let collabFile`

Reply:
Yeah it still needs to run in both cases

Reply:
In the `new Y.Map()` case we're having to create the new `Y.Array` and insert that into the map, and then insert all of the top level elements into that. I guess we could duplicate the code that inserts the elements, but I don't think that would necessarily make the code any clearer.